pages/index.tsx

```
@@ -0,0 +1,173 @@
+import React, { useEffect, useRef, useState } from "react";
+import fetch from "node-fetch";
+import * as jwt from "jsonwebtoken";
+import {
+  CLIENT_ID,
+  HOSTED_UI_LOGIN_URL,
+  HOSTED_UI_TOKEN_URL,
+  REDIRECT_URI,
+} from "./api/constants";
+import {
+  DecodedIdToken,
+  IdDetails,
+  SsoDetails,
+  TokenResponse,
+} from "./api/types";
+import generateSaml from "./api/generateSaml";
+import CallCenterPicker from "../components/CallCenterPicker";
+import SelfSubmittingSsoForm from "../components/SelfSubmittingSsoForm";
+
+export default function Home() {
+  const [idDetails, setIdDetails] = useState<IdDetails | undefined>(undefined);
+  const [ssoDetails, setSsoDetails] = useState<SsoDetails | undefined>(
+    undefined,
+  );
+  const [pickerSubmitWasPressed, setPickerSubmitWasPressed] = useState(false);
+  const [singleSubmitTriggered, setSingleSubmitTriggered] = useState(false);
+  const [error, setError] = useState<string | undefined>(undefined);
+  // Control double effect-invocation when `fetch` should only happen once
+  const tokenFetchAlreadyRan = useRef<boolean>(false);
+
+  useEffect(() => {
+    // ensure `document` is going to be defined
+    if (typeof window !== "undefined") {
+      const authorizationCode = (
+        document.location.search
+          .substring(1)
+          .split("&")
+          .find((s) => s.match(/^code=/)) ?? ""
+      ).split("=")[1];
+
+      // No authorizationCode? Go directly to the Hosted UI login page
+      if (authorizationCode == null || authorizationCode === "") {
+        window.location.replace(HOSTED_UI_LOGIN_URL);
+      } else {
+        if (!tokenFetchAlreadyRan.current) {
+          tokenFetchAlreadyRan.current = true;
+          fetch(HOSTED_UI_TOKEN_URL, {
+            method: "POST",
+            headers: { "Content-Type": "application/x-www-form-urlencoded" },
+            body: new URLSearchParams({
+              grant_type: "authorization_code",
+              client_id: CLIENT_ID!,
```

why does `CLIENT_ID!` need a `!` here?

---

`URLSearchParams` requires that the arguments here all be defined, but `CLIENT_ID` has type `string | undefined`.

This is not a very clean solution, I suppose if `CLIENT_ID` actually was undefined then I should throw an error, but it is unclear where and how I should throw that (ideally it happens during startup), and this approach was something we did in various parts of the UI webapp code, so I thought I could get away with it o.O

---

ah i see. maybe the `!` could be in `api/constants`, so it's type is never `undefined` after that?

`export const CLIENT_ID = process.env.NEXT_PUBLIC_COGNITO_CLIENT_ID;`

---

please see commit "ok fine" 😝